pkg/kubelet/network/hostport/hostport.go

```diff
@@ -140,32 +145,127 @@ func portMappingToHostport(portMapping *PortMapping) hostport {
 	}
 }
 
-// ensureKubeHostportChains ensures the KUBE-HOSTPORTS chain is setup correctly
-func ensureKubeHostportChains(iptables utiliptables.Interface, natInterfaceName string) error {
+// ensureKubeHostportChainLinked ensures the KUBE-HOSTPORTS chain is linked into the root of iptables
+func ensureKubeHostportChainLinked(iptables utiliptables.Interface) error {
 	glog.V(4).Info("Ensuring kubelet hostport chains")
-	// Ensure kubeHostportChain
+	// Ensure kubeHostportsChain exists
 	if _, err := iptables.EnsureChain(utiliptables.TableNAT, kubeHostportsChain); err != nil {
 		return fmt.Errorf("Failed to ensure that %s chain %s exists: %v", utiliptables.TableNAT, kubeHostportsChain, err)
 	}
+
+	// Enusure and fill kubeLocalDestChain with current info
+	if err := syncLocaldest(iptables); err != nil {
+		return fmt.Errorf("Failed to populate %s chain %s: %v", utiliptables.TableNAT, kubeLocalDestChain, err)
+	}
+
+	// Ensure it is linked from the root.
 	tableChainsNeedJumpServices := []struct {
 		table utiliptables.Table
 		chain utiliptables.Chain
 	}{
 		{utiliptables.TableNAT, utiliptables.ChainOutput},
 		{utiliptables.TableNAT, utiliptables.ChainPrerouting},
 	}
-	args := []string{"-m", "comment", "--comment", "kube hostport portals",
+	args := []string{
+		"-m", "comment", "--comment", "maybe kube hostport",
 		"-m", "addrtype", "--dst-type", "LOCAL",
-		"-j", string(kubeHostportsChain)}
+		"-j", string(kubeLocalDestChain),
+	}
 	for _, tc := range tableChainsNeedJumpServices {
 		if _, err := iptables.EnsureRule(utiliptables.Prepend, tc.table, tc.chain, args...); err != nil {
 			return fmt.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", tc.table, tc.chain, kubeHostportsChain, err)
 		}
+		removeOldLink(iptables, tc.table, tc.chain) // ignore errors
+	}
+	removeOldSNAT(iptables) // ignore errors
+
+	return nil
+}
+
+func syncLocaldest(iptables utiliptables.Interface) error {
+	start := time.Now()
+	defer func() {
+		glog.V(4).Infof("syncLocaldest took %v", time.Since(start))
+	}()
+
+	// Get iptables-save output so we can check for existing chains and rules.
+	// This will be a map of chain name to chain with rules as stored in iptables-save/iptables-restore
+	existingNATChains := make(map[utiliptables.Chain]string)
+	iptablesSaveRaw, err := iptables.Save(utiliptables.TableNAT)
```

Comment:
the only benefit of `iptables-save` here is to preserve the stats of kubeLocalDestChain? Can we ignore it?